File: src/MinCamlSharp.Tests/Parser/ParserTests.cs
using MinCamlSharp.CodeModel;
using MinCamlSharp.CodeModel.Tokens;
using MinCamlSharp.Parser;
using NUnit.Framework;

namespace MinCamlSharp.Tests.Parser
{
	[TestFixture]
	public class ParserTests
	{
		[Test]
		public void CanParseSumFunction()
		{
			// Arrange.
			var lexer = new Lexer(null, @"let rec sum x =
				if x <= 0 then 0 else
				sum (x - 1) + x in");
			var parser = new MinCamlSharp.Parser.Parser(null, lexer.GetTokens());

			// Act.
			var compilationUnitNode = parser.Parse();

			// Assert.
			//Assert.AreEqual("basic_material", fragment.Name);
			//Assert.IsNotNull(fragment.Parameters);
			//Assert.AreEqual(2, fragment.Parameters.VariableDeclarations.Count);
			//Assert.AreEqual(TokenType.Float, fragment.Parameters.VariableDeclarations[0].DataType);
			//Assert.AreEqual("alpha", fragment.Parameters.VariableDeclarations[0].Name);
			//Assert.AreEqual("ALPHA", fragment.Parameters.VariableDeclarations[0].Semantic);
			//Assert.AreEqual(TokenType.Float3, fragment.Parameters.VariableDeclarations[1].DataType);
			//Assert.AreEqual("color", fragment.Parameters.VariableDeclarations[1].Name);
		}
	}
}

File: src/MinCamlSharp.Compiler.Tests/ParserTests.fs
module MinCamlSharp.Compiler.Tests.Parser

open MinCamlSharp.Compiler
open MinCamlSharp.Compiler.Syntax
open NUnit.Framework
open NUnit.Framework.Constraints

[<Test>]
let CanParseSumFunction() =
    // Arrange.
    let code =  @"
        let rec sum x =
            if x <= 0 then 0 else
            sum (x - 1) + x in
        print_int (sum 10000)"
    let lexBuffer = Lexer.CreateBufferFromString(code)

    // Act.
    let tree = Parser.exp Lexer.token lexBuffer

    // Assert.
    let expectedTree =
        LetRec (
          { name = ("sum", Type.Var(ref None));
            args = [("x", Type.Var(ref None))];
            body = If(
                       LE(Var("x"), Int(0)),
                       Int(0),
                       Add(App(Var("sum"), [Sub(Var("x"), Int(1))]), Var("x"))
                     )},
            App (
                Var("print_int"),
                [App(Var("sum"), [Int(10000)])]
            )
        )
    Assert.That(tree, Is.EqualTo(expectedTree))